src/x11/i3/ipc/connection.hpp

```diff
@@ -0,0 +1,133 @@
+#include <qlocalsocket.h>
+#include <qobject.h>
+#include <qqml.h>
+#include <qqmlintegration.h>
+#include <qtmetamacros.h>
+#include <qtypes.h>
+
+#include "../../../core/model.hpp"
+#include "../../../core/qmlscreen.hpp"
+
+namespace qs::i3::ipc {
+
+class I3Workspace;
+class I3Monitor;
+} // namespace qs::i3::ipc
+
+Q_DECLARE_OPAQUE_POINTER(qs::i3::ipc::I3Workspace*);
+Q_DECLARE_OPAQUE_POINTER(qs::i3::ipc::I3Monitor*);
+
+namespace qs::i3::ipc {
+
+constexpr std::string MAGIC = "i3-ipc";
+
+enum EventCode {
+	RUN_COMMAND = 0,    // NOLINT
+	GET_WORKSPACES = 1, // NOLINT
+	SUBSCRIBE = 2,
+	GET_OUTPUTS = 3, // NOLINT
+	GET_TREE = 4,    // NOLINT
+
+	WORKSPACE = 0x80000000,
+	OUTPUT = 0x80000001,
+	UNKNOWN = 999,
+};
+
+using Event = std::tuple<EventCode, QJsonDocument>;
+
+///! I3/Sway IPC Events
+/// Emitted by @@I3.rawEvent(s)
+class I3IpcEvent: public QObject {
+	Q_OBJECT;
+
+	/// The name of the event
+	Q_PROPERTY(QString type READ type CONSTANT);
+	/// The payload of the event in JSON format.
+	Q_PROPERTY(QString data READ data CONSTANT);
+
+	QML_NAMED_ELEMENT(I3Event);
+	QML_UNCREATABLE("I3IpcEvents cannot be created.");
+
+public:
+	I3IpcEvent(QObject* parent): QObject(parent) {}
+
+	[[nodiscard]] QString type() const;
+	[[nodiscard]] QString data() const;
+
+	EventCode mCode = EventCode::UNKNOWN;
+	QJsonDocument mData;
+
+	static EventCode intToEvent(uint32_t raw);
+	static QString eventToString(EventCode event);
+};
+
+class I3Ipc: public QObject {
+	Q_OBJECT;
+
+public:
+	static I3Ipc* instance();
+
+	[[nodiscard]] QString socketPath() const;
+
+	QVector<Event> makeRequest(const QByteArray& request);
+	void dispatch(const QString& payload);
+
+	static QByteArray buildRequestMessage(EventCode cmd, const QByteArray& payload = QByteArray());
+	I3Workspace* findWorkspaceByName(const QString& name);
+	I3Monitor* findMonitorByName(const QString& name);
+
+	void setFocusedWorkspace(I3Workspace* workspace);
+	void setFocusedMonitor(I3Monitor* monitor);
+
+	void refreshWorkspaces();
+	void refreshMonitors();
+
+	I3Monitor* monitorFor(QuickshellScreenInfo* screen);
+
+	[[nodiscard]] I3Monitor* focusedMonitor() const;
+	[[nodiscard]] I3Workspace* focusedWorkspace() const;
+	[[nodiscard]] ObjectModel<I3Monitor>* monitors();
+	[[nodiscard]] ObjectModel<I3Workspace>* workspaces();
+signals:
+	void connected();
+	void rawEvent(I3IpcEvent* event);
+	void focusedWorkspaceChanged();
+	void focusedMonitorChanged();
+
+private slots:
+	void eventSocketError(QLocalSocket::LocalSocketError error) const;
+	void eventSocketStateChanged(QLocalSocket::LocalSocketState state);
+	void eventSocketReady();
+
+	void onFocusedWorkspaceDestroyed();
+	void onFocusedMonitorDestroyed();
+
+private:
+	explicit I3Ipc();
+
+	void onEvent(I3IpcEvent* event);
+
+	void handleWorkspaceEvent(I3IpcEvent* event);
+
+	static QVector<std::tuple<EventCode, QJsonDocument>> parseResponse(QByteArray rawEvent);
+
+	QLocalSocket liveEventSocket;
+	QLocalSocket commandSocket;
```

Comment:
Didn't find a way to circumvent a race condition coming from the live ipc event sending a message while still waiting for the command response, since I am returning the reply to the QS user

Reply:
If you're returning all events and not just requested ones you can use a single socket since you'd be racing anyway. Otherwise 2 is fine

Reply:
I am emitting all events through rawEvent, but dispatch returns to the user the IPC's reply to the requested command

Reply:
Sorry, haven't had time to go over this in detail before now, will comment on the larger issue in main review.